### ISSUE TITLE
Flow 4.0.0 doesn't retain binary compatibility with Flow 3.x and older

### ISSUE THREAD

**Issue by mvysny** [vaadin/flow#8831]
Having an utility jar library with a bunch of Vaadin components which calls `select.setItems(Collection)` will throw `java.lang.NoSuchMethodError: com.vaadin.flow.component.select.Select.setItems(Ljava/util/Collection;)V` since the function is now in `HasListDataView`.

If the library needs to target both Vaadin 16 and Vaadin 17, it either needs to be recompiled for both targets, or call `setItems()` via reflection.

**Comment by jhult** [comment]
Some additional details can be found in [vaadin-on-kotlin#51](https://github.com/mvysny/vaadin-on-kotlin/issues/51).

**Comment by Artur-** [comment]
Is this really a bug or is it this the reason for the version being 4 and not 3.x ?

**Comment by mvysny** [comment]
I was hoping to open a discussion about possible ways to retain the binary compatibility (e.g. by having missing methods - `setItems()`  and possibly others - explicitly defined on all components that previously implemented `HasDataProvider`), or whether we're okay with breaking the binary compatibility between Vaadin 16-17 (then documenting that in release notes).

**Comment by Legioth** [comment]
The rationale for breaking backwards compatibility is elaborated towards the end of https://vaadin.com/blog/data-binding-to-grid-gets-easier-and-more-efficient.

**Comment by mvysny** [comment]
@Legioth fair point. Still, the contract of `setItems()` hasn't changed between `HasDataProvider` and newer interfaces - wouldn't it be a gesture of good will to at least fix the binary incompatibility for `setItems()`?

**Comment by Legioth** [comment]
The contract has changed, the old method returns `void` and the new method returns a `DataView` instance.

Please let us know if there are ideas on how to maintain backwards compatibility for the compiler without breaking backwards compatibility in coders' minds (i.e. without introducing new concepts to learn about).

**Comment by mvysny** [comment]
Still, in the particular case of `setItems()`, change of return type is much smaller change than having `java.lang.NoSuchMethodError` thrown.

**Comment by pleku** [comment]
Unless I'm missing something, the changes are intended and while open discussion is good - we've had these discussions extensively internally and also had public tickets about the changes. While initially we wanted to avoid breaking changes and would have introduced a new API, in the end we decided to break things for a minority of users so that the majority of the users would have the same API as before.

The bad part is that the list of breaking changes is not visible in https://github.com/vaadin/flow/releases/tag/4.0.0.beta1 and that is my fault. We'll fix that.

**Comment by Legioth** [comment]
> Still, in the particular case of `setItems()`, change of return type is much smaller change than having `java.lang.NoSuchMethodError` thrown.

The JVM links methods based on the full signature, including the return type. As a consequence, `java.lang.NoSuchMethodError` is (to my understanding) the inevitable consequence of changing a return type and not recompiling code that invokes that method.

**Comment by pleku** [comment]
The breaking changes are now written in the release notes like they should have originally. While the discussion is welcome, I think we have not yet seen such feedback that we would consider reverting the API design to non breaking one which would have meant a 3.x version instead.

I'll keep this issue open for now to see if more feedback is added.

**Comment by mstahv** [comment]
We did a short investigation using add-ons and guessed around 3 add-ons would need to be recompiled or make minor code changes. https://docs.google.com/document/d/1BMzEpS1b-HTp1tXfCN1RqZVAFGFyR6dz41QnH0xh9V0/edit?usp=sharing Since that there are also some changes in the interfaces, but we expect those also to affect just a couple of libraries such as vaadin-on-kotlin and viritin.

We'll try to learn more about this and see if/how we can backport the changes later to the V14 branch. All reports for both directions are more than welcome!
